astropy/io/fits/file.py

```diff
@@ -76,6 +80,17 @@
 else:
     HAS_PATHLIB = True
 
+def _normalize_fits_mode(mode):
+    if mode is not None and mode not in IO_FITS_MODES:
+        if TEXT_RE.match(mode):
+            raise ValueError(
+                "Text mode '{}' not supported: "
+                "files must be opened in binary mode".format(mode))
+        new_mode = FILE_MODES.get(mode)
+        if new_mode not in IO_FITS_MODES:
+            raise ValueError("Mode '{}' not recognized".format(mode))
```

Comment:
In order to properly handle all types of file objects, the default cannot actually be `readonly` in all cases. If a path or URL is passed, then the effective default is still `readonly`, so nothing has changed. However, if a file object is passed, we want the default to match the mode of the given file object. For example, if I open a file with `'wb'` and pass the handle to `fitsopen`, I don't want my default to be `readonly`, I want it to be `ostream`.

With respect to the file mode check in `_open_fileobj`, that may be an oversight on my part but I need to take a closer look. It's likely that that check is now redundant and should be removed. In any case, that check does not happen early enough to catch erroneous cases where the mode of the given `fileobj` does not match the requested mode.

I attempted to thoroughly test all of these combinations and interactions in the new unit tests (which were added since the last time this was reviewed). However, it's possible that I missed some edge cases, in which case I'd be happy to add more tests.

Reply:
> However, if a file object is passed, we want the default to match the mode of the given file object. For example, if I open a file with 'wb' and pass the handle to fitsopen, I don't want my default to be readonly, I want it to be ostream.

Currently the behavior is that the file object mode must match the "fits mode", not the opposite. If the file mode does not match the fits mode (with the test in `_open_fileobj`) then an error is raised. 
And the fits mode is determined from the file mode only if fits mode is `None`.

So with your PR and comment cited above I'm not sure if you try to invert this logic, and why this would be needed for the original purpose, reading compressed file objects. Maybe it's just that the check in `_open_fileobj` was missing a few cases, or that check should be done earlier ?

Reply:
Here's the basic problem and an outline of the process I had to go through to make the compressed files work.: 
* if we use 'readonly' as the default, then we don't know if the user explicitly provided 'readonly' as the FITS mode, or whether no FITS mode was provided and we needed to choose a default. 
* It turns out that 'readonly' is not a valid default in all cases: if a file handle is provided, the default depends on the mode that was used to open that file handle.
* Once I fixed this, I realized that agreement between FITS mode and the underlying file mode was not being checked properly in all cases
* Once I made the check work properly, the check in `_open_fileobj` became redundant

If the FITS default mode handling and mode checking had been working properly in the first place, then this PR would have been **much** simpler 😅

Reply:
Thanks for details @drdavella ! I will also try to explain better what bothers me with this change :wink: 

> Here's the basic problem and an outline of the process I had to go through to make the compressed files work.:
>   - If we use 'readonly' as the default, then we don't know if the user explicitly provided 'readonly' as the FITS mode, or whether no FITS mode was provided and we needed to choose a default. This matters when we pass a compressed file handle.

There is no need "to choose a default", if the file mode is not compatible with the FITS mode (*) an exception must be raised. That's the current behavior (for file handles) and I don't see why it should be different with a *compressed* file handle ?

(*) - just to be clear on terminology, I mean the `mode` argument for `fitsopen` ('readonly' etc.)

>   - It turns out that 'readonly' is not a valid default in all cases: if a file handle is provided, the default depends on the mode that was used to open that file handle.

Nope, the file mode must be in agreement with the FITS mode. That's the root of the issue for me: with your changes, the file handle mode becomes more important than the FITS mode. And the FITS mode becomes useless (for file handles) as you set it to None by default and guess it from the file mode.

Also, I'm not saying that this way to proceed is wrong, but it may have unexpected consequences on existing codes (even if the probability is weak, when dealing with file modes you can have surprises), and this change seems unnecessary with respect to the purpose of the PR. 

Maybe I'm wrong here, but from what I understand the meaning of the `mode` argument for `fitsopen` is different now. For file handles it becomes a way to check the file handle mode. 

>   - Once I fixed this, I realized that agreement between FITS mode and the underlying file mode was not being checked properly in all cases
>   - Once I made the check work properly, the check in _open_fileobj became redundant

> If the FITS default mode handling and mode checking had been working properly in the first place, then this PR would have been much simpler :sweat_smile:

Sure, it's certainly possible that file mode agreement was not checked strictly enough, so fixing this is good!

Reply:
Okay, I think I understand your objection now. Will you accept changing the `mode` argument for `fitsopen` back to `'readonly'` as a solution? 

I will also need to remove one of the test cases in `astropy/io/fits/tests/test_core.py:test_open_file_handle` since it will no longer pass. However, according to your argument, the behavior it was testing was not expected to begin with. 

In particular, this tested whether it is possible to open a FITS file from a file handle that had been opened with `'wb'`. It sounds like you are saying that this should **not** be possible unless the `mode='ostream'` argument is explicitly provided to `fitsopen`. Is that correct?

Reply:
Yes exactly!